### ISSUE TITLE
$.inArray(void 0, ['foo', 'bar', void 0]) returns -1 (only in the real IE 8)

### ISSUE THREAD

**Issue by benjamin-albert** [jquery/jquery#2912]
`$.inArray(void 0, ['foo', 'bar', void 0])` will return `-1` when searching for `undefined` in an array that contains `undefined`.

As mentioned in the title, this problem cannot be reproduced in newer version of IE using IE 8 compatibility mode, only in the real IE 8.

![screen_shot_2016-02-09_at_9 42 25_am](https://cloud.githubusercontent.com/assets/5201428/12925659/fdabfaba-cf67-11e5-9ff4-04bf9308a978.png)

The only other code we have running in the page is jQuery UI.

**Comment by KidSysco** [comment]
Here is where we got the test machine. 

https://dev.windows.com/en-us/microsoft-edge/tools/vms/mac/

**Comment by dmethvin** [comment]
Looks like your page is using jQuery 1.9.0 from two years ago. If the bug is present in 1.12.0 and you'd like to submit a patch we could consider it. Otherwise it won't be fixed since IE8 is no longer supported by Microsoft as of last month and jQuery 1.12.x patches, if they are needed, will only consist of low-risk regression fixes.

**Comment by benjamin-albert** [comment]
@dmethvin We also had this problem with version 1.11.3.

@KidSysco Can you try this with jQuery version 1.12.0?

**Comment by KidSysco** [comment]
![screen shot 2016-02-09 at 12 33 56 pm](https://cloud.githubusercontent.com/assets/2731230/12926521/71a26cfe-cf29-11e5-84b9-fedb13d67c72.png)

Same thing in 1.12.0.

**Comment by gibson042** [comment]
Our [sparse array detection](https://github.com/jquery/jquery/blob/1.12-stable/src/core.js#L371-L372) is the culprit... `undefined` values are indistinguishable from holes in IE8: http://jsbin.com/loyamebido

Should be an easy fix, though... don't skip holes when `elem` is `undefined` (technically a 1.x/2.x divergence, but IE8 is forcing our hand and sparse arrays are obviously broken anywhere needing that support anyway, plus the damage is confined to a rare scenario).

**Comment by mgol** [comment]
> technically a 1.x/2.x divergence

Why not fix it only in 3.0 then? I don't think we should diverge, especially that it's been that way for a long time and at this point we should only apply critical fixes to 1.12/2.2 branches.

**Comment by markelog** [comment]
> Why not fix it only in 3.0 then?

Because we don't support IE8 in 3.0 :)

**Comment by mgol** [comment]
Ah, lol, I didn't read carefully enough. Wouldn't changing that now be a breaking change, though? I imagine some people use jQuery 1.x even if they don't test on IE 8 and then treating undefined differently would break their code. This doesn't seem very safe to do in a patch release.

**Comment by markelog** [comment]
mmm, this works in all other browsers except in ie8, so it looks as a bug to me. But reaaaly edgy, since `void`? In Array?

**Comment by benjamin-albert** [comment]
As a side note I'm not sure I'd consider this edgy.

This function can be used as a succinct way to check if a value is equal to one of many values and sometimes one of them is `undefined`.

So instead of doing:

    var type = elem.prop('type');
    if (!elem.is('input') || (type !== 'text' && type !== 'month' && type !== void 0)) {
      // Error...
    }

We could just do:

    if (!elem.is('input') || $.inArray(elem.prop('type'), ['text', 'month', void 0]) < 0) {
      // Error...
    }

**Comment by markelog** [comment]
> As a side note I'm not sure I'd consider this edgy.

In five years or so, since culprit was added, it seems you are the first who reported about this, so i'm pretty confidant :)

**Comment by benjamin-albert** [comment]
*I am going to use this wonderful opportunity to rant about IE < 10 :)*

Now that I think about it, because this bug only happens in the **real IE 8**  the number of users searching for `undefined` in an array can probably be counted on one hand.

The only place I can find a real IE 8 (without having to download a 1+ GB VM) is in one of my companies old production servers (where no one ever opens a web browser on).

Besides, even if there is other code out there that searches for `undefined` in an array the user will most likely not be able to execute that code because:
* Other JavaScript in the page will throw the famous `Unspecified error`, because most people I know (including me) test using IE 8 emulation.

![](http://blog.tcs.de/images/2011-03-23-img01.jpeg)

* The page will not be laid out correctly (if at all).
* This browser is so slow the user will most likely give up before the browser finishes laying out and rendering the page.

Even if you fix this issue it doesn't change the fact that we have to workaround this in our plugin.

In fact I my self am not sure it's a good idea to introduce breaking changes for this.

@markelog Perhaps you can just mention this somewhere in the API docs or something.

**Comment by gibson042** [comment]
@mgol: IE8 is the only current case where `jQuery.inArray( undefined, [ undefined ] )` doesn't return `0`. We can fix that in 1.12.1, but at the cost of breaking `jQuery.inArray( undefined, new Array( 1 ) )`—also only in IE8. Since all other browsers will use native `indexOf`, this is purely a question of `jQuery.inArray` supporting either sparse arrays or `undefined` search elements _in that browser_. We have a request here for the latter, and fulfilling it won't even break a unit test AFAICT.

**Comment by mgol** [comment]
@gibson042 Thanks for the explanation, now I get it. I still think it's risky; we obviously won't have requests to keep current behavior before we change it, people only report something if they expect different behavior. OTH, I think we haven't received lots of requests to change this logic for IE 8. Considering that and the fact that we're just about to stop the development of the 1.x/2.x lines, it just sounds too risky for me to now tamper with it, especially in a patch release.

**Comment by dmethvin** [comment]
So yeah, it comes down to having this single bug report for IE8 only after several years. The reporter says they've worked around it in their own code. If we fix our side, we may break the opposite case and we don't know if *that* one exists in the wild, but it appears *this* one no longer does. Even companies running old legacy code in IE8 compat view on IE11 won't be encountering this problem.

**Comment by markelog** [comment]
Just let's get one thing straight, since it seems some of us are confused.

If we would fix this, without touching anything else and then decided to release, we would need to, according to the semver, release it in the `patch` version, if we bump minor or major version, that would  **violate** semver policy that we adhere.

I think we shouldn't invent new policies for the releases. Otherwise any and every change could be considered as breaking, which is absurd.

**Comment by mgol** [comment]
> If we would fix this, without touching anything else and then decided to release, we would need to, according to the semver, release it in the patch version

I disagree. We'd fix one bug by introducing another one so it would be a breaking change.

**Comment by markelog** [comment]
> I disagree. We'd fix one bug by introducing another one so it would be a breaking change.

It is not a question of "agree" or "not agree, semver is pretty clear - bugfix is patch version

**Comment by scottgonzalez** [comment]
Why is this even still being discussed? After years of being in the wild, there has been one bug report. The reporter even says that if we fix it, it won't help them. The only affected environment isn't even supported anymore. I suggest we close the issue and move on.

**Comment by mgol** [comment]
> It is not a question of "agree" or "not agree", semver is pretty clear - bugfix is patch version

A bug fix that breaks the public API is not just a bug fix but a bug fix + a breaking change. And semver is, indeed, clear about what that means (http://semver.org/#spec-item-8):

8\. Major version X (X.y.z | X > 0) MUST be incremented if any backwards incompatible changes are introduced to the public API. It MAY include minor and patch level changes. Patch and minor version MUST be reset to 0 when major version is incremented.

But I agree with @scottgonzalez, we should just close this issue as wontfix.

**Comment by timmywil** [comment]
@scottgonzalez Agreed. Let's leave this one alone.

**Comment by markelog** [comment]
> But I agree with @scottgonzalez, we should just close this issue as wontfix.

> Major version X (X.y.z | X > 0) MUST be incremented if any backwards incompatible

As i said, your interpretation suggest that every bugfix could be interpreted as back-incompat, since there  could be a code that depends on that bug.

> But I agree with @scottgonzalez, we should just close this issue as wontfix.

Same here
